Document tracing issues related to ingestion model (#382)
getsentry/develop: src/docs/sdk/research/performance/index.mdx

> The following two examples synthesize the scope propagation issues.

Author: Hmm here I meant a bit more than just "demonstrate". I used "synthesize" because, I my view at the time, the two examples fully cover and define the issues.

But perhaps I don't need such precision, and "demonstrate" is a simpler word, possibly easier to understand, without much loss in meaning.

> The following two examples synthesize the scope propagation issues.

Reviewer: Hmmm - I'm trying to come up with a better word than "demonstrate," then.

I replaced "synthesize" because it needs something or someone to be the subject, doing the synthesis of the examples/ideas/etc (the object) into some coherent whole. What about "encapsulate?"

> ### Cannot Determine Current Span

Author: Removed the numbers from headings because there are too many similar `<h3>` headings under "Span Ingestion Model", and it would be silly and error prone to manually number all of those.

If order would ever change, it would break anchors/links.

Author: Moved to https://github.com/getsentry/develop/pull/387

- [OpenTracing shim doesn't change context #2016](https://github.com/open-telemetry/opentelemetry-js/issues/2016)
- [Http Spans are not linked / does not set parent span #2333](https://github.com/open-telemetry/opentelemetry-js/issues/2333)

### Conflicting Data Propagation Expectations

> There is a conflict of expectations that appear whenever we add a `trace` function as discussed earlier, or simply try to address scope propagation with Zones.

Author: Moved to https://github.com/getsentry/develop/pull/387

> 

Author: I've applied these suggestions and moved all changes unrelated to "ingestion model" to a separate PR ⇒ #387. Thanks for taking the time to review the rest of the document!

Now we can use this example to compare and understand the difference between Sentry's span ingestion model and the model used by OpenTelemetry and other similar tracing systems.

In Sentry's span ingestion model, all spans that belong to a transaction must be sent all together in a single request. That means that all `B` spans must be kept in-memory for the whole duration of the `B*` transaction, including time spent on downstream services (the storage service in the example).

In OpenTelemetry's model, spans are batched together as they are finished and batches are sent as soon as either there is a certain number of spans in the batch or a certain time has passed. In our example, it could mean that the first 3 `B` spans are batched together and sent while the first `S*` transaction is still in progress in the storage service. Subsequently, other `B` spans are batched together and sent as they finish, until eventually the `B*` transaction span is also sent.

> While transactions are notably useful to group together spans and to explore operations of interest in Sentry, end users and instrumentation developers have an extra cognitive burden to understand and choose between a transaction or a span.

Author: Thanks for the two iterations. I think here we indeed need a third one, because the above two lost the distinction between Sentry user (also a developer) and Sentry SDK maintainer (people like me and you who write instrumentation code, but also other users who may do so).

> While transactions are notably useful to group together spans and to explore operations of interest in Sentry, end users and instrumentation developers have an extra cognitive burden to understand and choose between a transaction or a span.

Author: I went with this

```suggestion
While transactions are notably useful as a way to group together spans and to explore operations of interest in Sentry, the form in which they currently exist imposes extra cognitive burden. Both SDK maintainers and end users have to understand and choose between a transaction or a span when writing instrumentation code.
```

The issues that follow in the next few sections have been identified in the current ingestion model, and are all related to this dichotomy.

### Complex JSON Serialization of Transactions

In OpenTelemetry's model, all [spans follow the same logical format](https://github.com/open-telemetry/opentelemetry-proto/blob/ebef7c999f4dea62b5b033e92a221411c49c0966/opentelemetry/proto/trace/v1/trace.proto#L56-L235). Users and instrumentation libraries can provide more meaning to a span by attaching key-value attributes to any span. The wire protocol uses lists of spans to send data from one system to another.

Sentry's model, unlike OpenTelemetry's, makes a hard distinction between two types of span: transaction spans (often refered to as "transactions") and regular spans.

> In memory, transaction spans and regular spans have one distinction: transaction spans have one extra attribute, the transaction `name`.

Author: - Metadata only comes to life when the transaction is transformed into a "Transaction event" (unless you're thinking about something... specific to the JS implementation? If so, that's not part of the spec).
- The "span collectors" (I think you mean what we call "recorder" in SDKs) are internal state, not relevant here, because the focus is on essential attributes. Think that the "recorder" could live anywhere else without any change in behavior, it is an implementation detail. Contrast it to the `name` attribute, which is required.

> When serialized as JSON for ingestion, though, the differences are greater. Sentry SDKs must serialize regular spans to JSON in a format that directly resembles the in-memory spans. However, the serialization of a transaction span requires mapping its span attributes to a Sentry `Event` (originally used to report errors, expanded to fit new fields of exclusive use for transactions), while all child spans go as a list embedded in the `Event`.

Author: Took most of it, but made changes ending up with:

```suggestion
When serialized as JSON, though, the differences are greater. Sentry SDKs serialize regular spans to JSON in a format that directly resembles the in-memory spans. By contrast, the serialization of a transaction span requires mapping its span attributes to a Sentry `Event` (originally used to report errors, expanded with new fields exclusively used for transactions), with all child spans embedded as a list in the `Event`.
```

### Transaction Spans Gain Event Attributes

> When transactions are transformed from their in-memory representation to an `Event`, they gain more attributes that are not assignable to regular spans, such as `breadcrumbs`, `extra`, `contexts`, `event_id`, `fingerprint`, `release`, `environment`, `user`, etc.

Author: Hmm I've never seen "etc" followed by double punctuation, did you really mean that?!

> When transactions are transformed from their in-memory representation to an `Event`, they gain more attributes that are not assignable to regular spans, such as `breadcrumbs`, `extra`, `contexts`, `event_id`, `fingerprint`, `release`, `environment`, `user`, etc.

Author: I see others have asked about the subject in https://english.stackexchange.com/questions/8382/when-etc-is-at-the-end-of-a-phrase-do-you-place-a-period-after-it. The double period doesn't occur neither in develop.sentry.dev nor docs.sentry.io, so I'll take your suggestion except for the double period.

When the new type of event was introduced, `transaction`, it was soon decided that they would not go through the `BeforeSend` hook, for essentially two reasons:

- Prevent user code from relying on the dual form of transactions (sometimes looking like a span, sometimes like an event, see earlier sections above);
- Avoid existing `BeforeSend` functions that were written with only errors in mind from interfering with transactions, be it mutating them accidentally, dropping them altogether, or some other unexpected side-effect.

However, it was also clear that some form of lifecycle hook was necessary to allow users to do things like updating a transaction name.

We ended up with the middle ground of allowing the mutation/dropping of transaction events through `EventProcessor` (a more general form of `BeforeSend`), which solves problems by giving users immediate access to their data before it goes out of the SDK, but is also the worse of both worlds in that it exposes the transaction duality that was never intended to leak and is more complicated to use than `BeforeSend`.

> In OpenTelemetry, spans go through span processors, which are two lifecycle hooks: one when a span is started and one when it is ended.

Author: Not sure about this one. Span processors *are* by definition two hooks, they are not something on their own that gets "associated with" hooks.

> In OpenTelemetry, spans go through span processors, which are two lifecycle hooks: one when a span is started and one when it is ended.

Author: Went with 

```suggestion
By contrast, in OpenTelemetry spans go through span processors, which are two lifecycle hooks: one when a span is started and one when it is ended.
```

> In OpenTelemetry, spans go through span processors, which are two lifecycle hooks: one when a span is started and one when it is ended.

Reviewer: Okay - maybe what I think of as "hooks" are actually "events?" I always thought that the thing happening was the hook, and the associated function was what was hooked onto the hook... though I guess that's not quite right, either. Really there's the thing that happens (the event), the ability the framework has given you to trigger a function when that thing happens (the hook), and then the function that gets triggered (the... something) - and I'm thinking of the span processors as being in that last category, which is why to me they're not themselves the hooks.

Anyway, I think it's fine.

### Nested Transactions

> On one hand, Sentry's ingestion model does not allow for nested transactions. Transactions are meant to mark service transitions.
> 
> On the other hand, Sentry as a product requires that transactions exist. Data outside of a transaction is lost.

Author: Thanks for noticing the confusion. Reworded as

```suggestion
Sentry's ingestion model was not designed for nested transactions within a service. Transactions were meant to mark service transitions.
```

> In practice, SDKs have no way of preventing transactions from becoming nested. The end result is possibly surprising to users, as each transaction starts a new tree. The only way to relate those trees is through the `trace_id`.

Author: > 1. When you're talking about spans being "nested," it seems like what you really mean is "simultaneous." To my mind, they're not nested unless they actually _are_ correctly associated with each other by using the same trace id.

Spans were "made to be nested". Transactions were not (made to be nested in the same service).

By "nested" I mean the two transactions have an ancestor-descendant relationship and are produced in the same service. Note that in the one-transaction-per-service model, as in the example in the beginning of the section, transactions are obvious nested within a trace. 

> 2\. I'm not sure how much this matters, but IIRC SDKs have different behavior here. For example, in JS if you have transaction A on the scope and you replace it with transaction B, transaction A is lost unless there's something else referencing it, because when B ends it just stays there, and even if it does get popped off, A never gets put back on. In Python, OTOH, context managers do pop off B, and they then put A back.

^^ this shows how, in different ways, "nested transactions" are not well-supported. Still, there is nothing stopping anyone from calling `startTransaction -> startChild -> startTransaction (oops) -> ...`

> Sentry's billing model is per event/transaction. That means that a transaction within a transaction generates two billable events.

Author: In reality billing is more complicated than that, but this document doesn't need the details.

> In SDKs, a transaction within a transaction causes inner spans to be "swallowed" by the innermost transaction surrounding them. In some situations, automatic instrumentation creating spans will only make them appear in one of the two transactions, but not both, causing apparent gaps in instrumentation.

Author: This is only the case because 1) the SDKs split spans by transaction (and they must do it) and 2) the UI (at the moment) only shows a single transaction at a time, ignoring anything that may exist around it in the same trace.

> The product does not know how to deal with nested transactions yet. If the transactions share a `trace_id` they can be visualized in the trace view. To get to the trace view, one must follow a link from one of the transactions. However, when looking at any one transaction it is as if all the others did not exist (either ancestor or descendent transactions are not represented on screen).

Author: The trace view shows the transactions, but only the transactions (without their spans).

I think the point here is that the system was not designed for the use case of nesting transactions within a service. Everything is built around the idea that each service contributes one transaction to a trace.

> The product does not know how to deal with nested transactions yet. If the transactions share a `trace_id` they can be visualized in the trace view. To get to the trace view, one must follow a link from one of the transactions. However, when looking at any one transaction it is as if all the others did not exist (either ancestor or descendent transactions are not represented on screen).

Author: Rewritten as

```suggestion
Sentry's UI is not designed to deal with nested transactions in a useful way. When looking at any one transaction it is as if all the other transactions in the trace did not exist (no other transaction is directly represented on the tree view). There is a trace view feature to visualize all transactions that share a `trace_id`, but the trace view only gives an overview of the trace by showing transactions and no child spans. There is no way to navigate to the trace view without first visiting some transaction.
```

There is also user confusion about what one would expect in the UI for such a situation (pseudocode):

```python
# if do_a_database_query returns 10 results, is the user
#   - seeing 11 transactions in the UI?
#   - billed for 11 transactions?
#   - see spans within create_thumbnail in the innermost transaction only?
with transaction("index-page"):
    results = do_a_database_query()
    for result in results:
        if result["needs_thumbnail"]:
            with transaction("create-thumbnail", {"resource": result["id"]}):
                create_thumbnail(result)
```

### Spans Cannot Exist Outside of a Transaction

> Sentry's tracing experience is centered entirely around the part of a trace that exists inside transactions. This means that data cannot exist outside of a transaction even if it exists in a trace.

Author: A trace is the set of all spans that share a `trace_id`.

The current model forces us to have transactions (because they are the spans that can be transformed into events for ingestion), but they are not fundamental (and not part of other systems).

In theory, there can be traces with no transactions whatsoever.

> If the SDK does not have a transaction going, a manual instrumentation by the user with a regular span is entirely lost. This is less of a concern on instrumented web servers, where a transaction is live most of the time, because automatic instrumentation starts and finishes a transaction for every incoming request.

Author: Note to self: `s/a user's manual//`

> If the SDK does not have a transaction going, a manual instrumentation by the user with a regular span is entirely lost. This is less of a concern on instrumented web servers, where a transaction is live most of the time, because automatic instrumentation starts and finishes a transaction for every incoming request.

Author: Rewritten as

```suggestion
If the SDK does not have a transaction going, regular spans created by instrumentation are entirely lost. That said, this is less of a concern on web servers, where automatically instrumented transactions start and finish with every incoming request.
```

> The requirement of a transaction is specially challenging on frontends (browser, mobile, and desktop applications), because in those cases auto-instrumented transactions are less reliable as they typically only last for some time before being automatically finished.

Author: ```suggestion
The requirement of a transaction is especially challenging on frontends (browser, mobile, and desktop applications), because in those cases auto-instrumented transactions less reliably capture all spans, as they only last for a limited time before being automatically finished.
```

> In our [example trace](#span-ingestion-model), the first span that originates the trace is due to a button click. If the button click `F*` was instrumented as a regular `span`, most likely no data from the frontend would be captured, but the other `B` and `S` spans would still be captured.

Author: This sounds good to me, thanks!

> In Sentry's model, if a span is not a transaction and there is no ancestor span that is a transaction, then the span won't be ingested. This, in turn, means there are many situations where a trace is missing crucial information that can help debug issues, particularly on the frontend where transactions need to end at some point but execution might continue.

Author: > The span won't even be sent, let alone ingested, right?

Correct. In many places in the document I use "ingest" as the arrow between SDK and Sentry/Relay. Not trying to be precise about sending/receiving/processing/validating/accepting/storing/etc.

> Execution of what?

Execution of the browser tab.

> Automatic and manual instrumentation have a challenge deciding whether to start a span or a transaction, and the decision is particularly difficult considering that poor [scope propagation](#scope-propagation) limits the existence of concurrent transactions.

Author: I politely disagree. Our current single-global-hub for browser doesn't allow for concurrent transactions. Would you have an example to show the contrary?

I can only imagine concurrent transactions using exclusively manual instrumentation and awkward constructs, not the obvious way the SDK was meant to be used.

If you have a short example, we could include it in this document (and even in docs.sentry.io) to show how to do concurrent transactions.

> Automatic and manual instrumentation have a challenge deciding whether to start a span or a transaction, and the decision is particularly difficult considering that poor [scope propagation](#scope-propagation) limits the existence of concurrent transactions.

Author: Let me clarify, I disagree more with myself than with you, @lobsterkatie ;)
And I agree with your edits.

Reading this sentence a second time, I realized that this section is not about concurrency -- it should not end talking about "concurrent transactions".

Rewritten as

```suggestion
Automatic and manual instrumentation have the challenge of deciding whether to start a span or a transaction, and the decision is particularly difficult considering that:
- If there is no transaction, then the span is lost.
- If there is already a transaction, then there is the [nested transactions](#nested-transactions) issue.
```

This causes transactions to be missing some Web Vitals or to have non-final measurements, for example for the LCP measurement.

### Unreliable Frontend Transaction Duration

Because all data must go in a transaction. Sentry's browser SDK creates a transaction for every page load and every navigation. Those transactions must end at some time.

If the browser tab is closed before the transaction is finished and sent to Sentry, all collected data is lost. Therefore, the SDK needs to balance the risk of losing all data and capturing as much useful data as possible.

Transactions are finished after a certain idle time after the last activity has been observed (for example, outgoing HTTP requests). This means that the duration of the page load / navigation transaction is a rather arbitrary value that can't necessarily be compared or improved, as it doesn't indicate any concrete and undestandable duration.

We counter the limitation by focusing on the LCP Web Vital as the default performance metric for browser, but, as discussed above, the final LCP value might be lost.

### In-memory Buffering Affects Servers

> Any system that may operate with several concurrent transactions at one time, which is the case of web servers, will be required to buffer complete span trees in memory.

Author: I'll rewrite this.

> Any system that may operate with several concurrent transactions at one time, which is the case of web servers, will be required to buffer complete span trees in memory.

Author: Rewritten

```suggestion
As discussed earlier, the current ingestion model requires Sentry SDKs to observe complete span trees in memory. Applications that operate with a constant flow of concurrent transactions will require considerable system resources to collect and process tracing data. Web servers are the typical case that exhibit this problem.
```

> This means recording 100% of spans and 100% of transactions for many server-side applications, even in a simplified form, is not feasible due to the overhead incurred.

Author: Rewritten

```suggestion
This means that recording 100% of spans and 100% of transactions is not feasible for many server-side applications, because the overhead incurred is just too high.
```

### Lack of Batch Ingestion

> When multiple concurrent transactions finish nearly at the same time, SDKs must send as many requests as there are transactions. There is no provision to batching multiple transactions into a single request.
> 
> This inneficiency means at least extra bandwidth consumption for frontends (browser, mobile) and extra overhead for backends.

Author: Thanks! I took another shot and rewrote as:

```suggestion
Sentry's ingestion model does not support ingesting multiple events at once. In particular, SDKs cannot batch multiple transactions into a single request.

As a result, when multiple transactions finish at nearly the same time, SDKs are required to make a separate request for each transaction. This behavior is at best highly inefficient and at worst a significant and problematic drain on resources such as network bandwidth and CPU cycles.
```

### Compatibility

> The special treatment of transactions is incompatible with OpenTelemetry which means we cannot implement an OpenTelemetry Exporter that can feed data into Sentry (though we have a [Sentry Exporter with a major correctness limitation](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/sentryexporter#known-limitations)). Likewise we cannot leverage OpenTelemetry SDKs and instrumentations.

Reviewer: Rearranged so the parenthetical doesn't break up the main idea.

```suggestion
The special treatment of transactions makes us incompatible with OpenTelemetry - we can't implement an OpenTelemetry Exporter that can feed data into Sentry, nor can we leverage OpenTelemetry SDKs and instrumentations. (Though we have an existing [Sentry Exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/sentryexporter#known-limitations), it suffers from major correctness limitations.)
```

Author: Thanks. I took the inspiration and rewrote as

```suggestion
The special treatment of transaction spans is incompatible with OpenTelemetry. Users with existing applications instrumented with OpenTelemetry SDKs cannot easily use Sentry to ingest and analyze their data.

Sentry does provide a Sentry Exporter for the OpenTelemetry Collector, but, due to the current ingestion model, [the Sentry Exporter has a major correctness limitation](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/sentryexporter#known-limitations).
```

## Summary

> We have learned a lot with the current tracing implementation at Sentry. This document is an attempt to capture many of the known limitations to serve as basis for future improvement.

Reviewer: ```suggestion
We have learned a lot through building the current tracing implementation at Sentry. This document is an attempt to capture many of the known limitations, in order to serve as the basis for future improvement.
```

> Tracing is a complex subject, and taming that complexity is no easy feat.

Reviewer: I really like this sentence. I feel like I should steal it for the top of the nextjs SDK docs. "Next.js is a complex framework, and taming that complexity is no easy feat. We're working on it - be patient."

> The first group of issues related to **scope propagation** is a concern exclusive to SDKs and how they are designed. Addressing that will require internal architecture changes to all SDKs, including the redesign of old features like breadcrumbs. Those would be pre-requisites to implementing simple-to-use tracing helpers like a `trace` function that works in any context and captures accurate and reliable performance data.
> 
> Note that such changes would almost necessarily mean new major versions of SDKs that break compatibility with existing versions.

Reviewer: I made the issues in the group - rather than the group itself - the subject of the first sentence, to avoid the correct-but-awkward-sounding "single group, plural issues, single is" construction.

I also felt like the "Note that..." sentence had a hard time standing on its own as a full paragraph. I'm guessing you did that to emphasize the point, though, so I italicized it.

```suggestion
Issues in the first group - those related to **scope propagation** - are a concern exclusive to SDKs and how they are designed. Addressing them will require internal architecture changes to all SDKs, including the redesign of old features like breadcrumbs, but making such changes is a prerequisite for implementing simple-to-use tracing helpers like a `trace` function that works in any context and captures accurate and reliable performance data. Note that such changes _would almost certainly mean releasing new major versions of SDKs that break compatibility with existing versions_.
```

Author: Took the suggestion without the italics in the last sentence.

> The second group of issues related to the **span ingestion model** is a lot more complex as it affects more parts of the product and requires a coordinated effort from multiple teams to change.

Reviewer: Same as above re: single, plural, single.

```suggestion
Issues in the second group - those related to the **span ingestion model** - are a lot more complex, as any changes made to solve them would affect more parts of the product and require a coordinated effort from multiple teams.
```

Author: Applied literally, thanks.

> Nonetheless, changes to the ingestion model would have an immeasurable impact on the product, as it could reduce the cognitive burden of instrumentation, collect more data, improve efficiency.

Reviewer 2: ```suggestion
Nonetheless, changes to the ingestion model would have an immeasurable impact on the product, as it could reduce the cognitive burden of instrumentation, collect more data, and improve efficiency.
```

> Nonetheless, changes to the ingestion model would have an immeasurable impact on the product, as it could reduce the cognitive burden of instrumentation, collect more data, and improve efficiency.

Reviewer: Reversed the order of the list in order to (dare I say it?) reduce the cognitive burden of reading the sentence. (As a rule of thumb, they say you're supposed to go from short/simple to long/complex when listing things, so that someone isn't trying to keep a complicated thing in their mind while also listening to new things.)

```suggestion
Nonetheless, making changes to the ingestion model would have an immeasurable, positive impact on the product, as doing so would improve efficiency, allow us to collect more data, and reduce the cognitive burden of instrumentation.
```

Author: I like the edit, thanks! Suggestion taken.

All those different expectations makes it hard to reuse, in an understandable way, the current notion of `scope`, how breadcrumbs are recorded, and how those different concepts interact.

> Finally, it is worth noting that the changes to restructure scope management most likely cannot be done without breaking existing SDK APIs. Existing SDK concepts, like hubs, scopes, breadcrumbs, user, tags, contexts, would all have to be remodeled.

Author: Additional note to the previous topic from #381, Scope Propagation.

Author: Moved to https://github.com/getsentry/develop/pull/387

All those different expectations makes it hard to reuse, in an understandable way, the current notion of `scope`, how breadcrumbs are recorded, and how those different concepts interact.

Finally, it is worth noting that the changes to restructure scope management most likely cannot be done without breaking existing SDK APIs. Existing SDK concepts, like hubs, scopes, breadcrumbs, user, tags, contexts, would all have to be remodeled.

## Span Ingestion Model

Consider a trace depicted by the following span tree:

```

> F

Author: Good that you mention... here I intended to play with the ambiguity, but didn't yet fully...
The "button click" as I described later is not typically a transaction.

In OTel's model, it would/could be just a regular span. In Sentry's model, if the span is not a transaction or there is no ancestor span that is a transaction, then it won't be ingested (but the other `B` and `S` spans would still be ingested!).

> F

Author: My original intention was just a hypothetical "trace", not necessarily one we could reproduce with Sentry.
But considering that 

> Spans that are at the entry point to a given service are marked with a `*` to denote that they are transactions.

The above is already a Sentry-specific requirement, then, yes, let's make this `F*`.

Pending a short item later about Frontend Transactions...

├─ B*
│  ├─ B
│  ├─ B
│  ├─ B
│  │  ├─ S*
│  │  ├─ S*
│  ├─ B
│  ├─ B
│  │  ├─ S*
│  ├─ B
│  ├─ B
│  ├─ B
│  │  ├─ S*

where
F: span created on frontend service
B: span created on backend service
S: span created on storage service
```

This trace illustrates 3 services instrumented such that a user clicks a button on a Web page (`F`) and a backend (`B`) performs some work which involves making several queries to a storage service (`S`). Spans that are at the entry point to a given service are marked with a `*` to denote that they are transactions.

Now we can use this example to compare and understand the difference between Sentry's span ingestion model and the model used by OpenTelemetry and other similar tracing systems.

> In Sentry's span ingestion model, all spans that belong to a transaction must be sent all together in a single request. That means that all `B` spans must be kept in-memory for the whole duration of the `B*` transaction, including time spent on downstream services (the storage service in the example).

Reviewer 2: Can we make it explicit that this is because span data cannot be sent to Sentry outside of a transaction. Feel like that can be worded better, but I want to make sure that people understand that transaction events -> only way to group and ingest performance data right now.

Author: Open to rewording suggestions :)

Author: I think this will be address by at least two of the problems related to the ingestion model:
- Spans Cannot Exist Outside of a Transaction
- Transaction Spans Must Be Converted to Transaction Events

The example should set the stage, and then we start listing out and exploring the identified issues.

Reviewer 2: > I think this will be addressed by at least two of the problems related to the ingestion model:

Agree, let's hold off on my suggestion then.

> In OpenTelemetry's model, spans are batched together as they are finished and batches are sent as soon as either there is a certain number of spans in the batch or a certain time has passed. In our example, it could mean that the first 3 `B` spans are batched together and sent while the first `S*` transaction is still in progress in the storage service. Subsequently, other `B` spans are batched together and sent as they finish, until eventually the `B*` transaction span is also sent.

Author: Without going into too much detail, there are basically two cases to send spans out of the process where the SDK runs:
- Timeout reached
- Batch size reached

https://github.com/open-telemetry/opentelemetry-go/blob/d57c5a5604851ed8c97ea6bbf4addff82cf009cf/sdk/trace/batch_span_processor.go#L91-L94

Of course there is flexibility for other strategies, but we don't really need to go over the basics. All examples in https://github.com/open-telemetry/opentelemetry-go/tree/main/example use a batch span processor.

The [built-in span processors as per spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#built-in-span-processors) are the Simple and Batch processors... the Simple is somewhat equivalent to a batch of always one.

> In OpenTelemetry's model, spans are batched together as they are finished and batches are sent as soon as either there is a certain number of spans in the batch or a certain time has passed. In our example, it could mean that the first 3 `B` spans are batched together and sent while the first `S*` transaction is still in progress in the storage service. Subsequently, other `B` spans are batched together and sent as they finish, until eventually the `B*` transaction span is also sent.

Author: BTW, regardless of SDK implementation, the OTLP wire protocol is based on sending a list of spans, grouped by instrumentation library, grouped by resource:

https://github.com/open-telemetry/opentelemetry-proto/blob/bacfe08d84e21fb2a779e302d12e8dfeb67e7b86/opentelemetry/proto/trace/v1/trace.proto#L27-L65

```proto
// A collection of InstrumentationLibrarySpans from a Resource.
message ResourceSpans {
  // The resource for the spans in this message.
  // If this field is not set then no resource info is known.
  opentelemetry.proto.resource.v1.Resource resource = 1;

  // A list of InstrumentationLibrarySpans that originate from a resource.
  repeated InstrumentationLibrarySpans instrumentation_library_spans = 2;
```

```proto
// A collection of Spans produced by an InstrumentationLibrary.
message InstrumentationLibrarySpans {
  // The instrumentation library information for the spans in this message.
  // Semantically when InstrumentationLibrary isn't set, it is equivalent with
  // an empty instrumentation library name (unknown).
  opentelemetry.proto.common.v1.InstrumentationLibrary instrumentation_library = 1;

  // A list of Spans that originate from an instrumentation library.
  repeated Span spans = 2;
```
